### DISCUSSION TITLE
Error handling differences in server.js and page.server.js

### DISCUSSION THREAD

**Discussion by NoelJacob** [sveltejs/kit#10205] [General]
So,
expected errors in server.js -> return data with error
unexpected errors in server.js -> handleError in hooks.server.js
expected errors in page.server.js -> show +error.js
unexpected errors in page.server.js -> handleError in hooks.server.js

Am I right? This should be a table in docs I guess because it'll save a lot of debugging time.

**Comment by teemingc** [comment]
You’re right. To put it simply: if an error occurs during page rendering (load function / page / layout), Kit will render the nearest `+error.svelte` page.

However, if the error occurs in `handle`, an endpoint, or the root layout `src/routes/+layout`, it will return the fallback page `src/error.html`. This static html page ensures no further errors will be thrown.

Endpoints return json if the request’s `accept` header asks for it.

**Reply by NoelJacob** [reply]
How do I call the handleError for expected errors from an endpoint? Because the error is being thrown from an async function that is not awaited inside endpoint and that might be running even after endpoint function returns.

**Reply by teemingc** [reply]
You'll have to replicate the error handling logic in the endpoint itself. The `handleError` function does not handle endpoint thrown errors.
